p2p/service.py

```diff
@@ -0,0 +1,53 @@
+import asyncio
+import logging  # noqa: F401
+
+from p2p.cancel_token import CancelToken
+from p2p.exceptions import OperationCancelled
+
+
+class BaseService:
+    # Subclasses should define their own logger.
+    logger = None  # type: logging.Logger
+    # Number of seconds stop() will wait for run() to finish.
+    _wait_until_finished_timeout = 2
+
+    def __init__(self, token: CancelToken) -> None:
+        self._finished = asyncio.Event()
+        self.cancel_token = token
+
+    async def run(self) -> None:
+        """Awaits for the subclass' _run() coroutine and sets the _finished event."""
+        try:
+            await self._run()
+        except OperationCancelled as e:
+            self.logger.info("%s finished: %s", self, e)
+        except Exception:
+            self.logger.exception("Unexpected error in %s, exiting", self)
+        finally:
+            self._finished.set()
+
+    async def stop(self) -> None:
```

Comment:
Should this raise an exception (or log a warning) when trying to stop a service that hasn't been started yet?

Reply:
I don't think that adds any value at this point, and it'd require another event/attribute to track whether or not run() has been called, so I'd rather not do that now